### PR TITLE
BUG: mtrand cannot be imported on Cygwin

### PR DESCRIPTION
**Pull request by trueroad**
There is an issue that mtrand in numpy-1.22.x cannot be imported
because `PyInit_mtrand` is not exported on Cygwin.
https://cygwin.com/pipermail/cygwin/2022-January/250481.html

mtrand is generated from three sources, mtrand.c,
src/legacy/legacy-distributions.c, and src/distributions/distributions.c.
In numpy-1.21.x, there are no symbols with
a `__declspec(dllexport)` attribute in the source files.
On the other hand, in numpy-1.22.0,
the commit b797a0cb97a8a9e7e36962dc8c37c075967aa88b adds the attribute
to the symbols in src/distributions/distributions.c.
`PyInit_mtrand` is contained in mtrand.c, so the attribute is not added.

Cygwin's ld exports all symbols if there are no symbols
with the `__declspec(dllexport)` attribute.
Therefore, in numpy-1.21.x,
`PyInit_mtrand` without the attribute is also exported.
However, if there is even one symbol
with the `__declspec(dllexport)` attribute,
Cygwin's ld does not export symbols without the attribute.
Therefore, in numpy-1.22.0, `PyInit_mtrand`
without the attribute is not also exported.
https://sourceware.org/binutils/docs/ld/WIN32.html

This commit solves the issue by removing the unnecessary
`__declspec(dllexport)` attribute.

### CODE REVIEW

**Comment by trueroad** [comment]
If I understand correctly, the cause of the CI failure seems to be due to the different Cython used.
The Cython I use is from the Cygwin package; the Cython used by the CI seems to be the one installed by pip.
https://github.com/numpy/numpy/runs/4906369348?check_suite_focus=true#step:8:48

Cygwin package's Cython has a patch that changes the macros for exported symbols.

```patch
--- origsrc/Cython-0.29.21/Cython/Utility/ModuleSetupCode.c	2020-07-08 23:44:39.000000000 +0200
+++ Cython/Utility/ModuleSetupCode.c	2021-01-30 08:34:37.402649500 +0100
@@ -709,7 +709,11 @@ static CYTHON_INLINE void * PyThread_tss
 /////////////// PyModInitFuncType.proto ///////////////
 
 #ifndef CYTHON_NO_PYINIT_EXPORT
-#define __Pyx_PyMODINIT_FUNC PyMODINIT_FUNC
+#ifdef __cplusplus
+#define __Pyx_PyMODINIT_FUNC extern "C" PyObject *
+#else
+#define __Pyx_PyMODINIT_FUNC PyObject *
+#endif
 
 #elif PY_MAJOR_VERSION < 3
 // Py2: define this to void manually because PyMODINIT_FUNC adds __declspec(dllexport) to it's definition.
```

Therefore, the `__declspec(dllexport)` attribute is not given to symbols by this Cython.
With this pull request, all symbols will no longer have the `__declspec(dllexport)` attribute and will be exported.

However, Cython installed using pip does not have the above patch applied.
In such an environment, the `__declspec(dllexport)` attribute seems to be given to the symbols.
Then, if you link .pyx and .c, the symbols in .pyx are exported, but the symbols in .c are not exported if they do not have the `__declspec(dllexport)` attribute.
Therefore, symbols that lost the `__declspec(dllexport)` attribute by this pull request no longer are exported, and the CI has failed.


Currently, the CI uses Cython installed with pip, but this may change to use Cygwin packaged Cython depending on versions or other conditions.
This will also change the conditions under which the error occurs.

**Comment by trueroad** [comment]
All CIs except ci/circleci are successful.
ci/circleci failed, but the cause seems to be #20877.
If I understand correctly, there is no problem with this pull request.

File: numpy/random/setup.py [hunk 1]
```diff
@@ -50,6 +50,13 @@ def generate_libraries(ext, build_dir):
         # Some bit generators require c99
         EXTRA_COMPILE_ARGS += ['-std=c99']
 
+    if sys.platform == 'cygwin':
+        # Export symbols without __declspec(dllexport) for using by cffi.
```
Comment (mattip) [inline numpy/random/setup.py:54]:
```suggestion
        # Export symbols without __declspec(dllexport) for using by cffi.
```
```suggestion
        # Export symbols without __declspec(dllexport) for using by cython.
```

**Comment by mattip** [comment]
Is this a general problem with cygwin + cython? Or are we doing something non-standard elsewhere?

File: numpy/random/setup.py (hunk 1 above)
Comment (trueroad) [reply numpy/random/setup.py]: Done. Thank you.

**Comment by trueroad** [comment]
> Is this a general problem with cygwin + cython?

I believe so.
It is due to the difference in behavior between Cygwin package's Cython and pip-installed Cython.

**Comment by mattip** [comment]
> It is due to the difference in behavior between Cygwin package's Cython and pip-installed Cython.

Is there some documentation about this in Cython that we can link to? An open issue or some other source for this?

**Comment by trueroad** [comment]
> Is there some documentation about this in Cython that we can link to? An open issue or some other source for this?

I am not aware of any such documentation, but the patch contained in the Cygwin package works that way.
I asked a question about the documentation and the behavior on the Cygwin mailing list.
Even if there is no such document, you can link to the discussion on this mailing list.
https://cygwin.com/pipermail/cygwin/2022-January/250599.html

**Comment by mattip** [comment]
It makes sense to me that NumPy will not be the only project with this problem. We can put this in as a temporary patch, but the cygwin-local cython patch should be submitted to cython.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
Let's give this a shot. Thanks @trueroad .

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @trueroad! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

**Comment by DWesl** [comment]
> > It is due to the difference in behavior between Cygwin package's Cython and pip-installed Cython.
> 
> Is there some documentation about this in Cython that we can link to? An open issue or some other source for this?

https://cygwin.com/pipermail/cygwin/2021-January/247216.html
seems to be the closest.

**Pull request merged**